test/xbgpu/test_engine.py

```diff
@@ -866,32 +917,69 @@ async def _send_data(
                     f"actual: {ig_recv['frequency'].value}."
                 )
 
-                corrprod_results[i][j] = ig_recv["xeng_raw"].value
+                corrprod_results[corrprod_output.name][j] = ig_recv["xeng_raw"].value
+
+        beam_results: dict[str, np.ndarray] = {}
+        if beam_capture_stop_heap_indices is not None:
+            # We adjust the corresponding beam stream to only receive the required
+            # amount of data (up until the capture-stop point).
+            for beam_output, capture_stop_heap_index in zip(beam_outputs, beam_capture_stop_heap_indices):
+                if capture_stop_heap_index > 0:
+                    # Obtaining a subset using zero (0) would result in an empty array.
+                    # We also don't want to use negative indices to work from the end
+                    # of the list of indices. Positive values are desired and more
+                    # explicit in this case.
+                    beam_results.update(
+                        {
+                            beam_output.name: np.zeros(
+                                (
+                                    len(batch_indices[:capture_stop_heap_index]),
+                                    n_channels_per_substream,
+                                    n_spectra_per_heap,
+                                    COMPLEX,
+                                ),
+                                bsend.SEND_DTYPE,
+                            )
+                        }
+                    )
+                else:
+                    beam_results.update(
+                        {
+                            beam_output.name: np.zeros(
+                                (len(batch_indices), n_channels_per_substream, n_spectra_per_heap, COMPLEX),
+                                bsend.SEND_DTYPE,
+                            )
+                        }
+                    )
+        else:
+            beam_results.update(
+                {
+                    beam_output.name: np.zeros(
+                        (len(batch_indices), n_channels_per_substream, n_spectra_per_heap, COMPLEX), bsend.SEND_DTYPE
+                    )
+                    for beam_output in beam_outputs
+                }
+            )
```

Comment:
This involves some repetition. It might be simpler to handle this case by instead starting with something like
```python
if beam_capture_stop_heap_indices is None:
    beam_capture_stop_heap_indices = [-1] * len(beam_outputs)
```
and then handling it with the same codepath as if it weren't None.